src/utils/include/utils/tensor.hpp

```diff
@@ -0,0 +1,194 @@
+
+    boost::transform(m_extents, std::back_inserter(m_strides),
+                     [stride = std::size_t(1)](std::size_t dim) mutable {
+                       auto const old = stride;
+                       stride *= dim;
+
+                       return old;
+                     });
+  }
+
+  /**
+   * @brief Element access.
+   * @param indices index for each dimension
+   */
+  template <typename... size_ts>
+  const_reference operator()(size_ts... indices) const {
+    return operator()({static_cast<std::size_t>(indices)...});
+  }
+
+  template <typename... size_ts> reference operator()(size_ts... indices) {
+    return operator()({static_cast<std::size_t>(indices)...});
+  }
+
+  template <typename Container>
+  const_reference operator()(Container const &indices) const {
+    assert(valid_indices(indices));
+    return m_data[calc_linear_index(indices)];
+  }
+
+  template <typename Container> reference operator()(Container const &indices) {
+    assert(valid_indices(indices));
+    return m_data[calc_linear_index(indices)];
+  }
+
+  const_reference
+  operator()(std::initializer_list<std::size_t> const &indices) const {
+    assert(valid_indices(indices));
+    return m_data[calc_linear_index(indices)];
+  }
+
+  reference operator()(std::initializer_list<std::size_t> const &indices) {
+    assert(valid_indices(indices));
+    return m_data[calc_linear_index(indices)];
+  }
+
+  /**
+   * @brief Bounds checked element access.
+   * @param indices index for each dimension
+   */
+  template <typename... size_ts> const_reference at(size_ts... indices) const {
+    return at({static_cast<std::size_t>(indices)...});
+  }
+
+  template <typename... size_ts> reference at(size_ts... indices) {
+    return at({static_cast<std::size_t>(indices)...});
+  }
+
+  template <typename Container>
+  const_reference at(Container const &indices) const {
+    validate_indices(indices);
+    return operator()(indices);
+  }
+
+  template <typename Container> reference at(Container const &indices) {
+    validate_indices(indices);
+    return operator()(indices);
+  }
+
+  const_reference at(std::initializer_list<std::size_t> const &indices) const {
+    validate_indices(indices);
+    return operator()(indices);
+  }
+
+  reference at(std::initializer_list<std::size_t> const &indices) {
+    validate_indices(indices);
+    return operator()(indices);
+  }
+
+  reference front() { return *begin(); }
+
+  const_reference front() const { return *cbegin(); }
+
+  reference back() { return m_data.back(); }
+
+  const_reference back() const { return m_data.back(); }
+
+  iterator begin() noexcept { return &(*m_data.begin()); };
+
+  const_iterator begin() const noexcept { return &(*m_data.begin()); };
+
+  const_iterator cbegin() const noexcept { return &(*m_data.begin()); };
+
+  iterator end() noexcept { return &(*m_data.end()); };
+
+  const_iterator end() const noexcept { return &(*m_data.end()); };
```

Comment:
You are not allowed to dereference the end pointer, it needs not to be valid, just say `data() + size()`, which is also less noisy.